src/server/engine_shard.cc

```diff
@@ -728,54 +729,78 @@ size_t EngineShard::CalculateEvictionBytes() {
   // state if flag has changed.
   bool rss_eviction_enabled_flag = GetFlag(FLAGS_enable_heartbeat_rss_eviction);
   if (eviction_state_.rss_eviction_enabled_ != rss_eviction_enabled_flag) {
-    eviction_state_.global_rss_memory_at_prev_eviction =
-        eviction_state_.deleted_bytes_before_rss_update = 0;
+    eviction_state_.shard_used_memory_at_prev_eviction =
+        eviction_state_.global_rss_memory_at_prev_eviction =
+            eviction_state_.acc_deleted_bytes_during_eviction =
+                eviction_state_.deleted_bytes_at_prev_eviction = 0;
     eviction_state_.rss_eviction_enabled_ = rss_eviction_enabled_flag;
   }
   if (eviction_state_.rss_eviction_enabled_) {
-    // Calculate how much rss memory is used by all shards
     const size_t global_used_rss_memory = rss_mem_current.load(memory_order_relaxed);
+    const size_t rss_memory_threshold_start = limit * (1. - eviction_memory_budget_threshold);
+    const size_t shard_used_memory = UsedMemory();
+
     auto& global_rss_memory_at_prev_eviction = eviction_state_.global_rss_memory_at_prev_eviction;
-    auto& deleted_bytes_before_rss_update = eviction_state_.deleted_bytes_before_rss_update;
+    auto& shard_used_memory_at_prev_eviction = eviction_state_.shard_used_memory_at_prev_eviction;
+    auto& deleted_bytes_at_prev_eviction = eviction_state_.deleted_bytes_at_prev_eviction;
+    auto& acc_deleted_bytes_during_eviction = eviction_state_.acc_deleted_bytes_during_eviction;
+
+    // Cap accumulated total deleted bytes
+    acc_deleted_bytes_during_eviction = std::min(
+        (global_used_rss_memory - rss_memory_threshold_start) / shards_count,
+        eviction_state_.acc_deleted_bytes_during_eviction + deleted_bytes_at_prev_eviction);
```

Comment:
I am confused about high-level algorithm here.

so for example, if `global_used_rss_memory - rss_memory_threshold_start) / shards_count < eviction_state_.acc_deleted_bytes_during_eviction + deleted_bytes_at_prev_eviction` 
you still decrease `acc_deleted_bytes_during_eviction -= deleted_bytes_at_prev_eviction;` 
which can again undeflow if say `global_used_rss_memory - rss_memory_threshold_start) / shards_count ` is less than `deleted_bytes_at_prev_eviction`. 
In general, it seems that the input here is `shard_rss_over_budget =  (global_used_rss_memory - rss_memory_threshold_start) / shards_count` and `shard_used_memory`.

So first step would be to factor out all these calculations into `eviction_state_.Update(shard_rss_over_budget,  shard_used_memory);`
and explain what `Update` does.